source/fuzz/transformation_inline_function.cpp

```diff
@@ -0,0 +1,239 @@
+  auto function_call_instruction_block =
+      ir_context->get_instr_block(function_call_instruction);
+  if (function_call_instruction == nullptr ||
+      function_call_instruction->opcode() != SpvOpFunctionCall ||
+      function_call_instruction !=
+          &*--function_call_instruction_block->tail() ||
+      function_call_instruction_block->terminator()->opcode() != SpvOpBranch) {
+    return false;
+  }
+
+  // The called function must not have an early return.
+  auto called_function = fuzzerutil::FindFunction(
+      ir_context, function_call_instruction->GetSingleWordInOperand(0));
+  if (called_function->HasEarlyReturn()) {
+    return false;
+  }
+
+  // |message_.result_id_map| must have an entry for every result id in the
+  // called function.
+  for (auto& block : *called_function) {
+    // Since the entry block label will not be inlined, only the remaining
+    // labels must have a corresponding value in the map.
+    if (&block != &*called_function->entry() &&
+        !message_.result_id_map().count(block.GetLabel()->result_id())) {
+      return false;
+    }
+    for (auto& instruction : block) {
+      // If |instruction| has result id then it must have a mapped id in
+      // |message_.result_id_map|.
+      if (instruction.HasResultId() &&
+          !message_.result_id_map().contains(instruction.result_id())) {
+        return false;
+      }
+    }
+  }
+
+  // The id mapping must not contain an entry for any parameter of the
+  // function that is being inlined.
+  bool found_entry_for_parameter = false;
+  called_function->ForEachParam(
+      [this, &found_entry_for_parameter](opt::Instruction* param) {
+        if (message_.result_id_map().contains(param->result_id())) {
+          found_entry_for_parameter = true;
+        }
+      });
+  return !found_entry_for_parameter;
+}
+
+void TransformationInlineFunction::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto function_call_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.function_call_id());
+  auto caller_function =
+      ir_context->get_instr_block(function_call_instruction)->GetParent();
+  auto called_function = fuzzerutil::FindFunction(
+      ir_context, function_call_instruction->GetSingleWordInOperand(0));
+
+  for (auto& block : *called_function) {
+    // The called function entry block label will not be inlined.
+    if (&block != &*called_function->entry()) {
+      auto cloned_label_instruction = block.GetLabelInst()->Clone(ir_context);
+      assert(message_.result_id_map().contains(
+                 cloned_label_instruction->result_id()) &&
+             "Label id must be mapped to a fresh id.");
+      uint32_t fresh_id =
+          message_.result_id_map().at(cloned_label_instruction->result_id());
+      cloned_label_instruction->SetResultId(fresh_id);
+      function_call_instruction->InsertBefore(
+          std::unique_ptr<opt::Instruction>(cloned_label_instruction));
```

Comment:
Actually, I don't think this will work. I think your tests might work because the functions always only have one basic block. But if you have additional basic blocks, you will need to add basic blocks, not just add label instructions. I _think_ adding a label instruction does not create a basic block automatically. See this example:

https://github.com/KhronosGroup/SPIRV-Tools/blob/150be20d4334b66c31a8ed81230f85b814b14a9f/source/fuzz/transformation_add_function.cpp#L276

Reply:
Yes, you would need to invalidate analyses to make IRContext aware of the new basic blocks. But in this case, you will potentially invalidate all pointers as well.

Reply:
Actually, I think invaliding analyzes is not even enough. I think the opt code base stores a module, not as a list of instructions, but as a list of functions, each with a list of basic blocks, each with a list of instructions. Thus, adding a label will not create a new basic block in the in-memory data structures, unless maybe you output the module as bytes and re-parsed it.

Reply:
Yes, right. OpLabels are not even stored in the instruction list. They are stored separately in each basic block.

Reply:
Done.